### ISSUE TITLE
Loading a module after workspace() hangs on 0.6+

### ISSUE THREAD

**Issue by ararslan** [JuliaLang/julia#22101]
I'm seeing this in 0.6-rc2 and current master:
```
              _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-DEV.371 (2017-05-27 13:28 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 71d14aeaec* (0 days old master)
|__/                   |  x86_64-apple-darwin16.6.0

julia> using Distributions

julia> workspace()

julia> using Distributions
```
This hangs seemingly indefinitely using 99.9% CPU. <kbd>Ctrl</kbd>+<kbd>C</kbd> does not interrupt the process; it must be forcibly killed.

It works on 0.5, it just emits a slew of method overwritten warnings rather than hanging.

**Comment by RalphAS** [comment]
Similar on Linux 0.6.0-rc2, except I was able to get a stacktrace showing recursion:

```julia
Stacktrace:
 [1] _include_from_serialized(::String) at ./loading.jl:157
 [2] _require_from_serialized(::Int64, ::Symbol, ::String, ::Bool) at ./loading.jl:200
 [3] _require_search_from_serialized(::Int64, ::Symbol, ::String, ::Bool) at ./loading.jl:236
 [4] _require(::Symbol) at ./loading.jl:434
 [5] require(::Symbol) at ./loading.jl:398
 [6] _include_from_serialized(::String) at ./loading.jl:157
 [7] _require_from_serialized(::Int64, ::Symbol, ::String, ::Bool) at ./loading.jl:200
 [8] _require_search_from_serialized(::Int64, ::Symbol, ::String, ::Bool) at ./loading.jl:236
 [9] _require(::Symbol) at ./loading.jl:434
 [10] require(::Symbol) at ./loading.jl:398

```

**Comment by vchuravy** [comment]
Should be fixed by #21714?

**Comment by ararslan** [comment]
Yes indeed. Thanks for the reference.
